ui/ozone/platform/wayland/wayland_input_method_context.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ui/ozone/platform/wayland/wayland_input_method_context.h"
+
+#include "base/bind.h"
+#include "base/memory/ptr_util.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "ui/events/event.h"
+#include "ui/ozone/platform/wayland/wayland_connection.h"
+#include "ui/ozone/platform/wayland/zwp_text_input_wrapper_v1.h"
+
+namespace ui {
+
+WaylandInputMethodContext::WaylandInputMethodContext(
+    WaylandConnection* connection)
+    : connection_(connection), text_input_(nullptr) {
+  use_ozone_wayland_vkb_ = getenv("USE_OZONE_WAYLAND_VKB");
+}
+
+WaylandInputMethodContext::~WaylandInputMethodContext() {
+  if (text_input_) {
+    text_input_->Deactivate();
+    text_input_->HideInputPanel();
+  }
+}
+
+void WaylandInputMethodContext::Initialize(
+    ui::mojom::LinuxInputMethodContextDelegatePtr delegate,
+    bool is_simple) {
+  delegate_ = std::move(delegate);
+  if (use_ozone_wayland_vkb_ && !is_simple && connection_->text_input_manager_v1()) {
+    text_input_ = std::make_unique<ZWPTextInputWrapperV1>(
+        connection_->text_input_manager_v1());
+    text_input_->Initialize(connection_, this);
+  }
+}
+
+void WaylandInputMethodContext::DispatchKeyEvent(
+    std::unique_ptr<ui::Event> key_event,
+    DispatchKeyEventCallback callback) {
+  const bool handled = false;
+  std::move(callback).Run(handled);
+}
+
+void WaylandInputMethodContext::Reset() {
+  if (text_input_)
+    text_input_->Reset();
+}
+
+void WaylandInputMethodContext::Focus() {
+  if (!text_input_ || !connection_->GetCurrentKeyboardFocusWindow())
+    return;
+
+  text_input_->Activate(connection_->GetCurrentKeyboardFocusWindow());
+  text_input_->ShowInputPanel();
+}
+
+void WaylandInputMethodContext::Blur() {
+  if (text_input_) {
+    text_input_->Deactivate();
+    text_input_->HideInputPanel();
+  }
+}
+
+void WaylandInputMethodContext::SetCursorLocation(const gfx::Rect& rect) {
+  if (text_input_)
+    text_input_->SetCursorRect(rect);
+}
+
+static base::string16 ConvertToUTF16(const std::string& text) {
+  if (base::IsStringUTF8(text)) {
```

Comment:
Why this? You should assume always UTF-8 on strings coming from text input interface.

Literally from specs: "      Text is generally UTF-8 encoded, indices and lengths are in bytes." https://cgit.freedesktop.org/wayland/wayland-protocols/tree/unstable/text-input/text-input-unstable-v1.xml#n41